### PR TITLE
Make file and crypto compatible and lots of other stuff

### PR DESCRIPTION
**Pull request by defuse**

### CODE REVIEW

**Comment by defuse** [comment]
-662 lines while retaining the same functionality, this might be my favorite commit ever.

**Comment by defuse** [comment]
TODO before merge: Fix outdated comments.

**Comment by paragonie-scott** [comment]
> -662 lines while retaining the same functionality, this might be my favorite commit ever.

I had a similar experience, converting Halite to PHP 7. Deleting code is wonderful.

**Comment by defuse** [comment]
Also, it's more likely than not that this introduces a serious security bug, so it needs a lot of review.

File: src/RuntimeTests.php [hunk 1]
```diff
@@ -228,7 +214,7 @@ private static function AESTestVector($config)
             throw new Ex\CryptoTestFailedException();
         }
 
-        $computed_plaintext = Crypto::plainDecrypt($ciphertext, $key, $iv, $config);
+        $computed_plaintext = Crypto::plainDecrypt($ciphertext, $key, $iv, Core::CIPHER_METHOD);
```
Comment (paragonie-scott) [inline src/RuntimeTests.php]: Instead of passing the constant, couldn't you just make that the default value for the parameter?

**Comment by paragonie-scott** [comment]
Other than a slight redundancy, :+1: LGTM

**Comment by defuse** [comment]
TODO before merge: Add tests for the salt string encoding (and tests that try to deserialize a key as a salt and vice-versa).

File: test/unit/BackwardsCompatibilityTest.php [hunk 2]
```diff
@@ -9,32 +9,6 @@ class BackwardsCompatibilityTest extends PHPUnit_Framework_TestCase
 {
     /**
      * @expectedException \Defuse\Crypto\Exception\InvalidCiphertextException
-     * @expectedExceptionMessage insecure fallback
-     */
-    function testDecryptLegacyWithWrongMethodWithHacks()
```
Comment (defuse) [inline test/unit/BackwardsCompatibilityTest.php]: Documenting why this test was removed: Because the `Core::LEGACY_VERSION` constant was removed.

**Comment by defuse** [comment]
This will close a lot of tickets: #124, #175, #194, #201, #202, #207, #209.

**Comment by paragonie-scott** [comment]
Merge conflicts? :(

**Comment by defuse** [comment]
@paragonie-scott: It's probably minor stuff caused by me merging random_compat, I'll rebase.

**Comment by defuse** [comment]
Superseded by #212.

**Pull request closed without merging**